File: 0622-design-circular-queue/solution.cpp
#include <vector>
using namespace std;

class MyCircularQueue {
private:
    vector<int> queue;
    int headIndex;
    int count;
    int capacity;

public:
    MyCircularQueue(int k) {
        capacity = k;
        queue.resize(k);
        headIndex = 0;
        count = 0;
    }
    
    bool enQueue(int value) {
        if (isFull()) {
            return false;
        }
        queue[(headIndex + count) % capacity] = value;
        count++;
        return true;
    }
    
    bool deQueue() {
        if (isEmpty()) {
            return false;
        }
        headIndex = (headIndex + 1) % capacity;
        count--;
        return true;
    }
    
    int Front() {
        if (isEmpty()) {
            return -1;
        }
        return queue[headIndex];
    }
    
    int Rear() {
        if (isEmpty()) {
            return -1;
        }
        return queue[(headIndex + count - 1) % capacity];
    }
    
    bool isEmpty() {
        return count == 0;
    }
    
    bool isFull() {
        return count == capacity;
    }
};

File: 0622-design-circular-queue/solution.py
class MyCircularQueue:
    def __init__(self, k):
        self.capacity = k
        self.queue = [0] * k
        self.headIndex = 0
        self.count = 0

    def enQueue(self, value):
        if self.isFull():
            return False
        self.queue[(self.headIndex + self.count) % self.capacity] = value
        self.count += 1
        return True

    def deQueue(self):
        if self.isEmpty():
            return False
        self.headIndex = (self.headIndex + 1) % self.capacity
        self.count -= 1
        return True

    def Front(self):
        if self.isEmpty():
            return -1
        return self.queue[self.headIndex]

    def Rear(self):
        if self.isEmpty():
            return -1
        return self.queue[(self.headIndex + self.count - 1) % self.capacity]

    def isEmpty(self):
        return self.count == 0

    def isFull(self):
        return self.count == self.capacity
